Reenable viewerCssTheme option for Firefox

hecerinc opened the pull request:
The `viewerCssTheme` was removed in #17222  and subsequently reenabled in #17293 , but only for Chromium and generic builds. This PR reenables this for Firefox as well.

Snuffleupagus commented:
Thanks for the patch, however there's a couple of overall issues with this:
 - Please note that we don't allow separate fixup commits in this project, hence you'll need squash those.
 - Please note that that you should never mix unrelated things in one PR, hence the link-file change needs to done separately.
 - Please make sure to write good commit messages, since a single line is rarely enough context. (Please take a look at the commit history to see how we usually write them.)
 - Finally, please note that we *purposely* removed `viewerCssTheme` in Firefox hence it's not entirely clear (at least to me) that we want to restore it.

hecerinc commented:
Thanks for the quick reply. I can remove the link-file change, no problem. 

I assumed the description for the PR is what would be merged when squash-committing the PR to the main branch, but I can write a better description if you think it's lacking. A lot of the commits in the history are one-liners, so I would ask for some pointers, please.

I understand the option was purposefully removed, but the original removal PR states that it was unstably implemented. It was then reenabled in #17293 with a different method, which is presumably more stable (as it's been checked in for a while), except it was not reenabled for Firefox.

I tried to ask about it repeatedly in that PR discussion (see https://github.com/mozilla/pdf.js/issues/17293#issuecomment-1872595105) to no response, so at the very least, I'd like this PR to be the catalyst for that conversation.

hecerinc commented:
@calixteman please weigh in here as to the reasoning behind the removal of the option

calixteman commented on web/app.js:
```diff
@@ -202,7 +202,7 @@ const PDFViewerApplication = {
     }
 
-    if (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) {
+    if (typeof PDFJSDev === "undefined") {
```
Did you check that it works as expected in Firefox ? (there is no chance it's working correctly)
You can compile Firefox in following the instructions here:
https://firefox-source-docs.mozilla.org/setup/index.html
Once the hg repo is cloned you can just run `./mach bootstrap` and select `Artifact mode` (except if you plan to hack on C++/rust code). 
You can then update the code of Firefox with your change in pdf.js in following what is done in (especially `gulp mozcentral`):
https://searchfox.org/mozilla-central/source/toolkit/components/pdfjs/update.sh
Once the above command is done you can just run `./mach build faster` et voila !

hecerinc replied:
I did check this, why would this not work?

Snuffleupagus replied:
Sorry, but did you *successfully* test changing the preference (via `about:config`) with this patch applied since it's unfortunately not possible for it to work!?

Furthermore, with the modified `PDFJSDev`-check above this functionality no longer works anywhere (i.e. not even in a GENERIC build)!

Snuffleupagus replied:
I'm guessing that the following is *actually* what you need here.
```diff
diff --git a/web/app.js b/web/app.js
index 7aa7596c6..c40e0b29b 100644
--- a/web/app.js
+++ b/web/app.js
@@ -202,39 +202,40 @@ const PDFViewerApplication = {
       await this._parseHashParams();
     }

-    if (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) {
-      let mode;
-      switch (AppOptions.get("viewerCssTheme")) {
-        case 1:
-          mode = "is-light";
-          break;
-        case 2:
-          mode = "is-dark";
-          break;
-      }
-      if (mode) {
-        document.documentElement.classList.add(mode);
+    let mode;
+    switch (AppOptions.get("viewerCssTheme")) {
+      case 1:
+        mode = "is-light";
+        break;
+      case 2:
+        mode = "is-dark";
+        break;
+    }
+    if (mode) {
+      document.documentElement.classList.add(mode);
+    }
+
+    if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
+      if (AppOptions.get("enableFakeMLManager")) {
+        this.mlManager =
+          MLManager.getFakeMLManager?.({
+            enableGuessAltText: AppOptions.get("enableGuessAltText"),
+            enableAltTextModelDownload: AppOptions.get(
+              "enableAltTextModelDownload"
+            ),
+          }) || null;
       }
-      if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
-        if (AppOptions.get("enableFakeMLManager")) {
-          this.mlManager =
-            MLManager.getFakeMLManager?.({
-              enableGuessAltText: AppOptions.get("enableGuessAltText"),
-              enableAltTextModelDownload: AppOptions.get(
-                "enableAltTextModelDownload"
-              ),
-            }) || null;
-        }
+    } else if (PDFJSDev.test("MOZCENTRAL")) {
+      if (AppOptions.get("enableAltText")) {
+        // We want to load the image-to-text AI engine as soon as possible.
+        this.mlManager = new MLManager({
+          enableGuessAltText: AppOptions.get("enableGuessAltText"),
+          enableAltTextModelDownload: AppOptions.get(
+            "enableAltTextModelDownload"
+          ),
+          altTextLearnMoreUrl: AppOptions.get("altTextLearnMoreUrl"),
+        });
       }
-    } else if (AppOptions.get("enableAltText")) {
-      // We want to load the image-to-text AI engine as soon as possible.
-      this.mlManager = new MLManager({
-        enableGuessAltText: AppOptions.get("enableGuessAltText"),
-        enableAltTextModelDownload: AppOptions.get(
-          "enableAltTextModelDownload"
-        ),
-        altTextLearnMoreUrl: AppOptions.get("altTextLearnMoreUrl"),
-      });
     }

     // Ensure that the `L10n`-instance has been initialized before creating
```

hecerinc replied:
Apologies! I must've overwritten a commit I had in one of the squashes, I had this change before :v

Snuffleupagus commented on web/app_options.js:
```diff
@@ -446,6 +446,11 @@ const defaultOptions = {
     value: 1,
     kind: OptionKind.API,
   },
+  viewerCssTheme: {
+    /** @type {number} */
+    value: typeof PDFJSDev !== "undefined" && PDFJSDev.test("CHROME") ? 2 : 0,
+    kind: OptionKind.VIEWER + OptionKind.PREFERENCE,
+  },
```
Please don't place a VIEWER option in the middle of API ones.

Snuffleupagus commented on web/app_options.js:
```diff
@@ -454,7 +459,6 @@ const defaultOptions = {
         : "../web/wasm/",
     kind: OptionKind.API,
   },
-
```
This line-break should not be removed, since it helps readability by separating API and WORKER options.

hecerinc committed:
Reenable viewerCssTheme option for Firefox

The `viewerCssTheme` was removed in #17222  and subsequently reenabled in #17293,
but only for Chromium and generic builds. This commit reenables the
function using the new method introduced in #17293.

hecerinc commented:
@Snuffleupagus  @calixteman  I have updated the PR based on your suggestions

calixteman reviewed (approved):
LGTM.
@Snuffleupagus is it fine for you ?

hecerinc commented:
Hi, thanks for approving :) are there any more steps needed to merge this?
